server/app/auth/oidc/applicant/OidcApplicantProfileAdapter.java

```diff
@@ -0,0 +1,168 @@
+package auth.oidc.applicant;
+
+import auth.CiviFormProfile;
+import auth.CiviFormProfileData;
+import auth.ProfileFactory;
+import auth.Roles;
+import auth.oidc.OidcProfileAdapter;
+import com.google.common.collect.ImmutableSet;
+import com.typesafe.config.Config;
+import java.util.Locale;
+import java.util.Objects;
+import javax.inject.Provider;
+import org.pac4j.core.credentials.Credentials;
+import org.pac4j.oidc.client.OidcClient;
+import org.pac4j.oidc.config.OidcConfiguration;
+import org.pac4j.oidc.profile.OidcProfile;
+import repository.UserRepository;
+
+/**
+ * This class ensures that the OidcProfileCreator that both the AD and IDCS clients use will
+ * generate a CiviFormProfile object. This is necessary for merging those accounts with existing
+ * accounts - that's not usually needed in web applications which is why we have to write this class
+ * - pac4j doesn't come with it. It's abstract because AD and IDCS need slightly different
+ * implementations of the two abstract methods.
+ */
+public abstract class OidcApplicantProfileAdapter extends OidcProfileAdapter {
+
+  protected final Config app_configuration;
+
+  protected final String localeAttributeConfigName = "locale_attribute";
+  protected final String firstNameAttributeConfigName = "first_name_attribute";
+  protected final String secondNameAttributeConfigName = "second_name_attribute";
+  protected final String emailAttributeConfigName = "email_attribute";
+
+  public OidcApplicantProfileAdapter(
+      OidcConfiguration configuration,
+      OidcClient client,
+      Config app_configuration,
```

Comment:
@Nullable, right? you're setting it to null in IdcsProfileAdapter.

Reply:
blocking: use Optional, avoid passing `null` whenever possible